File: bot/restart.py
"""Ciclo de vida: servidor opencode + processo do bot (/restart)."""
from pathlib import Path
import asyncio
import logging
import os
import signal
import subprocess
import time
from dotenv import dotenv_values
from . import config, state
from .turns import _kill_all_turns
from .opencode import oc_ensure_server


logger = logging.getLogger(__name__)


def _read_oc_port_from_env() -> tuple[int, str]:
    """Lê a porta/URL do opencode server direto do .env (sem mexer em os.environ),
    para que o /restart respeite uma mudança de porta feita em disco."""
    port, url = config.OC_PORT, config.OC_URL
    try:
        vals = dotenv_values(Path(__file__).resolve().parent.parent / ".env")
        raw = str(vals.get("OPENCODE_SERVER_PORT") or "").strip()
        if raw.isdigit():
            port = int(raw)
        url = str(vals.get("OPENCODE_SERVER_URL") or "").strip() or f"http://127.0.0.1:{port}"
    except Exception:
        pass
    return port, url


async def _find_opencode_pids(ports: set[int], uid: int) -> list[int]:
    """Retorna os PIDs (do usuário atual, excluindo o próprio processo) que
    correspondem a `opencode serve --port <porta>` para as portas dadas."""
    pids: list[int] = []
    for port in sorted(ports):
        proc = await asyncio.create_subprocess_exec(
            "pgrep", "-u", str(uid), "-f", f"opencode serve --port {port}",
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.DEVNULL,
        )
        out, _ = await proc.communicate()
        for x in out.decode().split():
            pid = int(x)
            if pid != os.getpid() and pid not in pids:
                pids.append(pid)
    return pids


def _pids_alive(pids: list[int]) -> list[int]:
    """Filtra `pids` mantendo só os que ainda existem (ou cujo estado não dá
    pra confirmar por falta de permissão — tratados como vivos por segurança)."""
    alive = []
    for pid in pids:
        try:
            os.kill(pid, 0)
            alive.append(pid)
        except ProcessLookupError:
            pass
        except PermissionError:
            alive.append(pid)
    return alive


async def _kill_opencode_servers(ports: set[int] | None = None, timeout: float = 15.0, max_rounds: int = 3) -> bool:
    """Derruba apenas os processos `opencode serve --port <porta>` do próprio
    usuário (incluindo órfãos da porta antiga), sem matar por padrão global
    outras instâncias do opencode na máquina. Retorna True se nenhum sobrou.

    Faz até `max_rounds` ciclos de SIGTERM -> espera -> SIGKILL antes de
    desistir. Isso é deliberadamente um loop com teto fixo (e não recursão
    infinita): se algum PID nunca puder ser confirmado como morto (ex.
    permissão negada, processo zumbi preso em D-state), o código antigo
    recursava sem limite a cada rodada e podia estourar o limite de
    recursão do Python numa máquina onde o processo simplesmente não morre.
    """
    ports = ports or {config.OC_PORT}
    uid = os.getuid()

    for round_num in range(1, max_rounds + 1):
        targets: list[int] = []
        for _ in range(3):
            targets = await _find_opencode_pids(ports, uid)
            if targets:
                break
            await asyncio.sleep(0.3)
        if not targets:
            return True

        for pid in targets:
            try:
                os.kill(pid, signal.SIGTERM)
            except (ProcessLookupError, PermissionError):
                pass

        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            if not _pids_alive(targets):
                return True
            await asyncio.sleep(0.25)

        for pid in targets:
            try:
                os.kill(pid, signal.SIGKILL)
            except (ProcessLookupError, PermissionError):
                pass
        await asyncio.sleep(0.5)

        still_alive = _pids_alive(targets)
        if not still_alive:
            return True
        if round_num == max_rounds:
            logger.error(
                "Não foi possível encerrar opencode serve nas portas %s após %d rodada(s); pids ainda vivos: %s",
                sorted(ports), max_rounds, still_alive,
            )
    return False


def _spawn_bot_process():
    """Relaça o bot num processo destacado, com o mesmo run.sh com que foi iniciado,
    para sobreviver ao encerramento do processo atual."""
    bot_dir = Path(__file__).parent
    log_path = bot_dir / "bot.log"
    log_fh = open(log_path, "ab", buffering=0)
    try:
        proc = subprocess.Popen(
            ["bash", "run.sh"],
            cwd=str(bot_dir),
            stdin=subprocess.DEVNULL,
            stdout=log_fh,
            stderr=subprocess.STDOUT,
            start_new_session=True,
            close_fds=True,
        )
        logger.info("Processo do bot relaçado (pid=%d)", proc.pid)
    finally:
        log_fh.close()


async def _cycle_server_locked() -> None:
    """Derruba o `opencode serve` atual (inclusive zumbis) e sobe um novo,
    recriando o cliente HTTP. O processo do bot continua o mesmo."""
    new_port, new_url = _read_oc_port_from_env()
    config.OC_PORT, config.OC_URL = new_port, new_url
    if state._client is not None:
        try:
            await state._client.aclose()
        except Exception:
            pass
        state._client = None
    await _kill_opencode_servers({config.OC_PORT})
    if state._server_proc is not None:
        try:
            await asyncio.wait_for(state._server_proc.wait(), timeout=5)
        except Exception:
            pass
        state._server_proc = None
    state._we_started_server = False
    await oc_ensure_server()
    logger.info("Servidor opencode reciclado (porta=%d)", config.OC_PORT)


async def _respawn_bot_and_exit() -> None:
    """Sobe um processo novo do bot via run.sh e encerra este via SIGTERM
    (o run_polling faz o shutdown gracioso). Quem chama deve antes desarmar
    o oc_stop_server (via state._we_started_server=False) se o servidor deve ficar no ar."""
    _spawn_bot_process()
    await asyncio.sleep(1.0)
    os.kill(os.getpid(), signal.SIGTERM)


async def _restart_server_only(chat_id: int):
    """Reinicia só o servidor opencode. O bot continua no ar."""
    app = state._app_ref
    try:
        await _kill_all_turns()
        await _cycle_server_locked()
        logger.info("Restart do servidor concluído (porta=%d)", config.OC_PORT)
        if app:
            try:
                await app.bot.send_message(
                    chat_id=chat_id,
                    text=f"✅ *Servidor reiniciado* — servidor `{config.OC_URL}` OK. O bot continuou no ar.",
                    parse_mode="Markdown",
                )
            except Exception:
                pass
    except Exception as e:
        logger.exception("Falha no restart do servidor")
        if app:
            try:
                msg = f"❌ *Falha no restart do servidor:* `{e}`\n\nVerifique o log: `.opencode_bot_server.log`"
                await app.bot.send_message(chat_id=chat_id, text=msg, parse_mode="Markdown")
            except Exception:
                pass


async def _restart_bot_only(chat_id: int):
    """Reinicia só o bot do Telegram (novo processo). O servidor continua no ar."""
    app = state._app_ref
    if app:
        try:
            await app.bot.send_message(
                chat_id=chat_id,
                text="🔁 *Reiniciando o bot...* volto em segundos. O servidor continua no ar.",
                parse_mode="Markdown",
            )
        except Exception:
            pass
    await _kill_all_turns()
    state._we_started_server = False  # post_shutdown não pode matar o servidor ao sair
    await _respawn_bot_and_exit()


async def _restart_both(chat_id: int):
    """Reinicia o servidor e o bot. O processo novo sobe um servidor fresco no boot."""
    app = state._app_ref
    if app:
        try:
            await app.bot.send_message(
                chat_id=chat_id,
                text="🔁 *Reiniciando bot + servidor...* volto em segundos.",
                parse_mode="Markdown",
            )
        except Exception:
            pass
    await _kill_all_turns()
    try:
        await _kill_opencode_servers({config.OC_PORT})
    except Exception:
        logger.exception("Falha ao derrubar o servidor no restart duplo")
    if state._server_proc is not None:
        try:
            await asyncio.wait_for(state._server_proc.wait(), timeout=5)
        except Exception:
            pass
        state._server_proc = None
    state._we_started_server = False
    await _respawn_bot_and_exit()


async def _perform_restart(chat_id: int, target: str = "both"):
    """Despacha o /restart: `bot`, `server` ou `both`."""
    if target == "bot":
        await _restart_bot_only(chat_id)
    elif target == "server":
        await _restart_server_only(chat_id)
    else:
        await _restart_both(chat_id)

File: src/restart.ts
/** Ciclo de vida: servidor opencode + processo do bot (/restart).
 * Port de `bot/restart.py`: alvos bot|server|both, kill por padrão de
 * processo com teto de rodadas, relançamento via run-hybrid.sh.
 */
import { type Bot } from "grammy";
import { endpoint, ensureServer, refreshEndpointFromEnv, stopServer } from "./opencode.ts";
import { killAllTurns } from "./turns.ts";

export type RestartTarget = "bot" | "server" | "both";

const ALIASES: Record<string, RestartTarget> = {
  bot: "bot",
  server: "server",
  servidor: "server",
  srv: "server",
  both: "both",
  ambos: "both",
  tudo: "both",
  all: "both",
};

export const RESTART_LABELS: Record<RestartTarget, string> = {
  bot: "o bot",
  server: "o servidor",
  both: "o bot + o servidor",
};

export function parseRestartTarget(arg: string | undefined): RestartTarget | null {
  if (!arg) return "both";
  return ALIASES[arg.trim().toLowerCase()] ?? null;
}

function serverPattern(port: number): string {
  return `opencode serve --port ${port}`;
}

function pidsOf(port: number): number[] {
  const args = ["pgrep", "-f", serverPattern(port)];
  try {
    const uid = typeof process.getuid === "function" ? process.getuid() : undefined;
    if (uid !== undefined) args.splice(1, 0, "-u", String(uid));
  } catch {
    /* sem getuid: pgrep sem filtro de usuário */
  }
  try {
    const out = Bun.spawnSync(args);
    return out.stdout
      .toString()
      .split(/\s+/)
      .map((x) => parseInt(x, 10))
      .filter((n) => Number.isFinite(n) && n !== process.pid);
  } catch {
    return [];
  }
}

function alive(pids: number[]): number[] {
  return pids.filter((pid) => {
    try {
      process.kill(pid, 0);
      return true;
    } catch (e) {
      const code = (e as NodeJS.ErrnoException)?.code;
      if (code === "EPERM") return true; // existe, sem permissão: trata como vivo
      return false;
    }
  });
}

/** Derruba `opencode serve --port` (com SIGTERM, depois SIGKILL, teto de
 * 3 rodadas). Retorna true se não sobrou nenhum. */
export async function killServers(ports?: Set<number>, timeoutMs = 15_000, maxRounds = 3): Promise<boolean> {
  const targets = ports ?? new Set([endpoint.port]);
  for (let round = 1; round <= maxRounds; round++) {
    let pids: number[] = [];
    for (let i = 0; i < 3; i++) {
      for (const p of targets) pids.push(...pidsOf(p));
      pids = [...new Set(pids)];
      if (pids.length) break;
      await Bun.sleep(300);
    }
    if (!pids.length) return true;
    for (const pid of pids) {
      try {
        process.kill(pid, "SIGTERM");
      } catch {
        /* já saiu */
      }
    }
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline) {
      if (!alive(pids).length) return true;
      await Bun.sleep(250);
    }
    for (const pid of pids) {
      try {
        process.kill(pid, "SIGKILL");
      } catch {
        /* já saiu */
      }
    }
    await Bun.sleep(500);
    const still = alive(pids);
    if (!still.length) return true;
    if (round === maxRounds) console.error(`restart: pids teimosos nas portas ${[...targets]}: ${still}`);
  }
  return false;
}

function respawnBotProcess(): void {
  const root = `${import.meta.dir}/..`;
  const proc = Bun.spawn(["bash", "run-hybrid.sh"], {
    cwd: root,
    stdin: "ignore",
    stdout: Bun.file(`${root}/bot.log`),
    stderr: Bun.file(`${root}/bot.log`),
    env: { ...process.env },
  });
  proc.unref();
  console.log(`restart: bot relançado (pid=${proc.pid})`);
}

async function respawnAndExit(bot: Bot): Promise<void> {
  respawnBotProcess();
  await Bun.sleep(1000);
  try {
    await bot.stop();
  } catch {
    /* polling já parado */
  }
  process.exit(0);
}

/** Recicla o `opencode serve` mantendo o processo do bot (port de _cycle_server_locked). */
export async function cycleServer(): Promise<void> {
  await refreshEndpointFromEnv();
  await stopServer();
  await killServers(new Set([endpoint.port]));
  await ensureServer();
  console.log(`restart: servidor reciclado (porta=${endpoint.port})`);
}

async function restartServerOnly(bot: Bot, chatId: number): Promise<void> {
  try {
    await killAllTurns(bot);
    await cycleServer();
    await bot.api.sendMessage(chatId, `✅ *Servidor reiniciado* — servidor \`${endpoint.url}\` OK. O bot continuou no ar.`, {
      parse_mode: "Markdown",
    }).catch(() => {});
  } catch (e) {
    console.error("restart do servidor falhou:", e);
    await bot.api.sendMessage(
      chatId,
      `❌ *Falha no restart do servidor:* \`${e}\`\n\nVerifique o log: \`.opencode_bot_server.log\``,
      { parse_mode: "Markdown" },
    ).catch(() => {});
  }
}

async function restartBotOnly(bot: Bot, chatId: number): Promise<void> {
  await bot.api.sendMessage(chatId, "🔁 *Reiniciando o bot...* volto em segundos. O servidor continua no ar.", {
    parse_mode: "Markdown",
  }).catch(() => {});
  await killAllTurns(bot);
  keepServerAliveOnExit();
  await respawnAndExit(bot);
}

async function restartBoth(bot: Bot, chatId: number): Promise<void> {
  await bot.api.sendMessage(chatId, "🔁 *Reiniciando bot + servidor...* volto em segundos.", {
    parse_mode: "Markdown",
  }).catch(() => {});
  await killAllTurns(bot);
  try {
    await killServers(new Set([endpoint.port]));
  } catch (e) {
    console.warn("restart duplo: falha ao derrubar servidor:", e);
  }
  await stopServer();
  keepServerAliveOnExit();
  await respawnAndExit(bot);
}

// O shutdown do index.ts mata o servidor que subimos; no restart do bot o
// servidor deve continuar no ar para o processo novo. Como o exit passa por
// outro caminho (process.exit direto), basta não chamar stopServer — este
// hook documenta a intenção e desarma o flag caso o shutdown rode no meio.
let exitKeepsServer = false;

export function keepServerAliveOnExit(): void {
  exitKeepsServer = true;
}

export function shouldKeepServerOnExit(): boolean {
  return exitKeepsServer;
}

/** Despacha o /restart: `bot`, `server` ou `both`. */
export async function performRestart(bot: Bot, chatId: number, target: RestartTarget): Promise<void> {
  if (target === "bot") await restartBotOnly(bot, chatId);
  else if (target === "server") await restartServerOnly(bot, chatId);
  else await restartBoth(bot, chatId);
}

export function restartConfirmKeyboard(): { text: string; data: string }[][] {
  return [
    [
      { text: "🤖 Só o bot", data: "__restart:bot" },
      { text: "🖥️ Só o servidor", data: "__restart:server" },
    ],
    [{ text: "🔁 Bot + servidor", data: "__restart:both" }],
    [{ text: "❌ Cancelar", data: "/status" }],
  ];
}
